lightning/src/offers/offer.rs

```diff
@@ -414,12 +414,23 @@ impl OfferContents {
 		self.amount.as_ref()
 	}
 
+	// Should only be called from InvoiceRequest since the Amount::Currency check in
+	// OfferContents::check_amount_msats_for_quantity will have already passed.
+	pub(super) fn amount_msats(&self) -> u64 {
+		match self.amount {
+			None => 0,
+			Some(Amount::Bitcoin { amount_msats }) => amount_msats,
+			Some(Amount::Currency { .. }) => unreachable!(),
```

Comment:
We still need to be able to parse a semantically valid offer. It's really that we can't create an `InvoiceRequest` (which shares `OfferContent`) from an `Offer` with a currency. I could make this return an `Option` or `Result`, but that just kicks the problem to the caller. Figured it would better to have this code near the code that checks for currency support.

Reply:
> I think we should just refuse to parse an offer if the amount is a currency. Parsing an offer and not being able to create an `InvoiceRequest` for it is basically useless, no? Let's just fix our data storage to be consistent with our assumptions so we don't have panics and unreachables lying around.

I don't think we want to limit LDK in that manner. Someone may want to use it, for instance, to create a BOLT 12 offer equivalent of https://lightningdecoder.com/.

> Alternatiely, we could check for the value and store it in the `InvoiceRequest`, which is probably what we'll want when we do add currency support anyway - store the msat conversion in `InvoiceRequest`. That will also avoid an unwrap/unreachable.

We need to support an `InvoiceRequest` without an amount set. Storing it would cause the value returned by `InvoiceRequestContents::as_tlv_stream` to be inaccurate, and thus break the signature calculation.

I can move the code back to `InvoiceRequest::amount_msats` if you prefer. Then at very least we can't accidentally call it from inside `Offer`.

Reply:
> Then we should be converting `Offer` into something else when we store it in an `InvoiceRequest`. Half the point of rust is that we can have a data model that matches our invariants so that we can avoid writing the invariants in code and we don't have to check things :)

To be fair, the only reason the invariant is there is because we don't want to support part of the spec yet. 😛 When we do support it, the code we be structured as is.

> 
> More generally, I'm a bit confused about how we're gonna _not_ check for currencies when decoding an `Offer`, but then we assume that no currency is present in `amount_msats`? That doesn't seem like an invariant for an `Offer`, so much as a check that should be in the `InvoiceRequest`?
> 

Right. The assumption is in `OfferContents::amount_msats`, which is only ever called from `InvoiceRequest`. I can move it back `InvoiceRequest::amount_msats` as I mentioned in my previous comment, which is how I originally had it (https://github.com/lightningdevkit/rust-lightning/pull/1927/commits/fd483ae6d42cfbd1c58a9adeae7094a899dec436).

> Huh? Why can't we store it in `InvoiceRequest` itself? We can ignore it or do anything we want with it when converting to a TLV stream. I assume I'm missing something about how amounts propagate.

Ah, I see what you're saying. We could do that at the cost of storing it twice and additional complexity. FWIW, this just moves the `unreachable` to the `InvoiceRequest` parsing code, IIUC.

Reply:
> I dont think so? I was imagining previously that the user would have to pass in a conversion when creating the `InvoiceRequestBuilder`, so we'd still have a new field there.

If passing a conversion when creating the builder, the builder would just set `amount_msats` on `InvoiceRequestContents`.

> I'm still confused by this - IIUC, by the time we have an `InvoiceRequest` it _must_ have an amount in msats. If an offer has no currency, we can't include an `amount_msats` in the `InvoiceRequest` when serialized, as the `invoiceRequest` has an implied amount. Thus, `InvoiceRequestContents` currently has an `Option` for the amount, as the amount may be stored in one of two places, leading to all this confusion.

At the serialization layer, the only time an invoice request MUST have an amount is when the offer doesn't specify an amount. Otherwise, it MAY be omitted according to the spec:

```
    - if `offer_amount` is not present:
      - MUST specify `invreq_amount`.
    - otherwise:
      - MAY omit `invreq_amount`.
      - if it sets `invreq_amount`:
        - MUST specify `invreq_amount`.`msat` as greater or equal to amount expected by `offer_amount` (and, if present, `offer_currency` and `invreq_quantity`).
```

That is, for the no currency case, it's still possible to have an amount. FWIW, there's a rationale given later for why a user may want to include it:

```
Users can give a tip (or obscure the amount sent) by specifying an
`invreq_amount` in their invoice request, even though the offer specifies an
`offer_amount`.  The recipient will only accept this if
the invoice request amount exceeds the amount it's expecting (i.e. its
`offer_amount` after any currency conversion, multiplied by `invreq_quantity`, if
any).
```

For the currency case, the same rules apply for the serialization layer -- the only time an amount MUST be included is when the offer doesn't have an amount. However, offers having a currency MUST also have an amount. Needing an `Option` in the invoice request is more about whether the requester decided to set an amount than the presence (or lack of) a currency in the offer. The requester MAY omit the amount even if the offer has a currency.

Anyhow, at the abstraction level, yes, an invoice request always has an amount -- either set explicitly or inferred from the fields mentioned in the spec. It just isn't determined by the currency, so that can't inform the serialization.

> If I'm right, here, ISTM we should make the `amount_msat` in `InvoiceRequest` non-`Option`al and handle this logic at the serialization layer. Then we wouldn't have any assumptions not captured by our data model, and this would upgrade seamlessly when we (eventually) support building `InvoiceRequest`s for an `Offer` with a currency.

Non-`Option`al in `InvoiceRequest` or `InvoiceRequestContents`? Just want to make sure I understood, though may be not relevant given my point above.

Reply:
> Both? Either?

There are implications of doing this that may not be all that desirable.

Since `OfferContents` has an optional `Amount`, `InvoiceRequestBuilder::new` would need to take an (optional?) amount for when that is `None`. Otherwise, it's unclear what the builder should initialize `InvoiceRequestContents::amount_msats` to. But when the offer _does_ have an amount, it would be strange to require that the user supply an amount when constructing the builder or even pass `None` if the parameter is an `Option` rather than them just explicitly calling `InvoiceRequestBuilder::amount_msats` only if they want to give a larger amount than necessary.

If instead it were only required on `InvoiceRequest` and optional on `InvoiceRequestContents`, we'd still need to set the amount on `InvoiceRequest` when parsing. But that still requires either having an `unreachable!()` in `InvoiceRequest::try_from` or needing to define `TryFrom` on `(InvoiceRequestContents, u64)` or something like that.

> ISTM we could make it non-optional in `InvoiceRequestContents` and then at the serialization layer we write one if its different from the one in the underlying `OfferContents` and omit it if its equal to the one there.

But why add this complexity -- and what would be required in the builder and parsing code mentioned above -- when we don't need to? Computing the amount when needed is no different from what we do when `chain` is not set. And the unreachable is only temporary. We can have a unit test that would fail when updating to ensure that it is removed.

Also note, for the currency case, during building we'd end up having the user pass in an amount in `new` but then we'd need to do some sort of check in `sign` to determine if `as_tlv_stream` should use `Some` or `None` for the amount. And that check would need to take into account the exchange rate, which may be different from when the builder is first constructed, unless we want to assume a constant exchange rate across the life of the builder.

I'd much prefer the simpler approach over this. I'm not opposed to considering a concrete patchset that you think would be acceptable. I'm just not seeing why all this the added complexity is desirable.

Reply:
> Won't it need to do that when we get to currency conversion anyway? We could have two methods, one which does not do currency conversion (and fails if no amount is passed and there isn't an amount in the offer) and one which takes an explicit amount and checks it?

A user only needs to set an amount if they want to pay more than the expected amount (or if the offer didn't contain an amount). AFAICT, the reader of the invoice request (i.e., payee) is the one that determines if this amount is sufficient when taking into account currency conversions.

So when the offer has an amount (currency or not), the normal case is to not set an amount in the invoice request. It's only when receiving an invoice does the payer check if they are happy with the exchange rate given by the payee.

> Because we need to to make the data model match the reality of what's going on to avoid unnecessary panics?

IIUC, the data model is "set an amount invoice request if you want to tip / obscure the amount" not "set an amount in invoice request that you are willing to pay". i.e., currency conversion only matters insomuch that you want to tip / obscure the amount of an offer with a currency denomination.

The panic is only there because we don't currently support currencies. It disappears once we do. In fact, the code in question is only needed when creating an invoice and thus could be moved entirely there without any changes to `InvoiceRequest` or `OfferContents`. Much of the discussion around the data model is resulting from the reuse of the "contents" types across higher-level structures.

> Having a check to figure out what to write in the serialization logic is way cleaner than adding additional preconditions we have to enforce at construction plus deserialization :).

I'm not sure I follow this point. Which additional preconditions?

> I'm confused, how is the unreachable temporary? We need it until we add an amount in the `InvoiceRequest{,Contents}` to support currencies.

Yeah, that's what I meant by temporary.

> I assume a builder only exists for a few seconds at max, not sure why we can't assume an exchange rate doesn't change during that time.
> 

Fair point. But note my point above about there shouldn't really need to care about any of this at `InvoiceRequest` creation time. It's only important after reading and deciding whether to respond with an `Invoice`.

> I'm not seeing what the "simpler approach" is here?

Leaving the code as is. 🙂

Or moving the checks into `InvoiceBuilder` rather than trying to "reuse" `InvoiceRequest` and `OfferContents`.

Reply:
Discussed offline. Since we only support non-currency offers, we can check the `Offer::amount` in `InvoiceBuilder::for_offer` and return an error if the `InvoiceRequest` is for an `Offer` with a currency. So I can drop the related commit and fixup from this PR.

The reasoning around this is that the payer does not need to care about conversion in the typical `InvoiceRequest` case. It's the payee who will need to convert to msats when building an `Invoice`. The payer should only need to deal with msats.

However, the payer's app may display the currency-to-msats conversion for the offer (and the opposite for the invoice) to the user before paying. We'll need to carefully document the payment flow so users are aware of how this is intended to work.

Reply:
Updated both PRs accordingly.